File: web/backend/engine.py
import discord
import aiohttp
import asyncio
import logging
import uuid
import datetime
from typing import Dict, List, Optional, Callable

class QueueHandler(logging.Handler):
    def __init__(self, callback: Callable[[str], None]):
        super().__init__()
        self.callback = callback

    def emit(self, record):
        try:
            msg = self.format(record)
            if asyncio.get_event_loop().is_running():
                asyncio.create_task(self.callback(msg))
        except Exception:
            self.handleError(record)

class MigrationEngine:
    def __init__(self, config: dict, log_callback: Callable[[str], None]):
        self.config = config
        self.log_callback = log_callback
        
        self.logger = logging.getLogger(f"engine_{uuid.uuid4().hex[:8]}")
        self.logger.setLevel(logging.INFO)
        # Avoid duplicate logs if engine is reused in same process
        self.logger.handlers = []
        handler = QueueHandler(log_callback)
        handler.setFormatter(logging.Formatter('%(asctime)s - %(levelname)s - %(message)s'))
        self.logger.addHandler(handler)
        
        self.role_map = {}
        self.channel_map = {}
        self.message_author_cache = {}

    async def stoat_request(self, token: str, method: str, path: str, json_data: dict = None) -> Optional[dict]:
        url = f"https://api.stoat.chat{path}"
        headers = {"X-Bot-Token": token}
        
        for attempt in range(3):
            async with aiohttp.ClientSession() as session:
                async with session.request(method, url, json=json_data, headers=headers) as resp:
                    if resp.status in [200, 201]:
                        return await resp.json()
                    elif resp.status == 429:
                        retry_after = (await resp.json()).get('retry_after', 1000) / 1000.0
                        self.logger.warning(f"Rate limited. Waiting {retry_after}s...")
                        await asyncio.sleep(retry_after + 0.1)
                        continue
                    else:
                        error = await resp.text()
                        self.logger.error(f"Stoat API Error ({method} {path}): {resp.status} - {error}")
                        return None
        return None

    def map_permissions(self, discord_perms: discord.Permissions) -> int:
        stoat_bits = 0
        if discord_perms.view_channel: stoat_bits |= (1 << 20)
        if discord_perms.send_messages: stoat_bits |= (1 << 22)
        if discord_perms.manage_messages: stoat_bits |= (1 << 24)
        if discord_perms.manage_channels: stoat_bits |= (1 << 2)
        if discord_perms.manage_roles: stoat_bits |= (1 << 3)
        if discord_perms.connect: stoat_bits |= (1 << 30)
        if discord_perms.speak: stoat_bits |= (1 << 31)
        return stoat_bits

    async def get_existing_structure(self, token: str, server_id: str):
        server = await self.stoat_request(token, "GET", f"/servers/{server_id}")
        if not server: return {}, {}
        roles = {r['name'].lower(): id for id, r in server.get('roles', {}).items()}
        channels_res = await self.stoat_request(token, "GET", f"/servers/{server_id}/channels")
        channels = {c['name'].lower(): c['_id'] for c in channels_res} if channels_res else {}
        return roles, channels

    async def run_clone(self, discord_token: str, stoat_token: str, source_id: int, target_id: str):
        self.logger.info(f"Starting cloning task for server {source_id} to {target_id}")
        client = discord.Client(intents=discord.Intents.default())
        
        @client.event
        async def on_ready():
            try:
                guild = client.get_guild(source_id)
                if not guild:
                    self.logger.error("Discord Guild not found.")
                    return

                existing_roles, existing_channels = await self.get_existing_structure(stoat_token, target_id)
                
                # 1. Roles
                for role in sorted(guild.roles, key=lambda r: r.position):
                    if role.is_default(): continue
                    name_lower = role.name.lower()
                    if name_lower in existing_roles:
                        self.role_map[role.id] = existing_roles[name_lower]
                        continue
                    
                    if self.config.get('dry_run'):
                        self.logger.info(f"[DRY RUN] Would create role: {role.name}")
                        self.role_map[role.id] = f"dry_role_{role.id}"
                        continue
                        
                    res = await self.stoat_request(stoat_token, "POST", f"/servers/{target_id}/roles", {
                        "name": role.name,
                        "permissions": [self.map_permissions(role.permissions), 0],
                        "colour": f"#{role.color.value:06x}" if role.color.value else None,
                        "hoist": role.hoist
                    })
                    if res and res.get('id'):
                        self.role_map[role.id] = res.get('id')
                        self.logger.info(f"Created role: {role.name}")

                # 2. Channels & Categories
                stoat_categories = []
                for category in sorted(guild.categories, key=lambda c: c.position):
                    cat_channels = []
                    for channel in sorted(category.channels, key=lambda ch: ch.position):
                        if not isinstance(channel, (discord.TextChannel, discord.VoiceChannel)): continue
                        
                        name_lower = channel.name.lower()
                        if name_lower in existing_channels:
                            chan_id = existing_channels[name_lower]
                            self.channel_map[channel.id] = chan_id
                            cat_channels.append(chan_id)
                            continue

                        if self.config.get('dry_run'):
                            self.logger.info(f"[DRY RUN] Would create channel: {channel.name} in {category.name}")
                            chan_id = f"dry_chan_{channel.id}"
                            self.channel_map[channel.id] = chan_id
                            cat_channels.append(chan_id)
                            continue

                        res = await self.stoat_request(stoat_token, "POST", f"/servers/{target_id}/channels", {
                            "name": channel.name,
                            "type": "Text" if isinstance(channel, discord.TextChannel) else "Voice",
                            "description": getattr(channel, 'topic', None)
                        })
                        if res and res.get('_id'):
                            chan_id = res.get('_id')
                            self.channel_map[channel.id] = chan_id
                            cat_channels.append(chan_id)
                            self.logger.info(f"Created channel: {channel.name}")

                    if cat_channels:
                        stoat_categories.append({"id": str(uuid.uuid4())[:8], "title": category.name, "channels": cat_channels})

                if stoat_categories and not self.config.get('dry_run'):
                    await self.stoat_request(stoat_token, "PATCH", f"/servers/{target_id}", {"categories": stoat_categories})

                self.logger.info("Server structure cloning finished.")
            except Exception as e:
                self.logger.error(f"Error in cloning: {e}")
            finally:
                await client.close()

        try:
            await client.start(discord_token)
        except Exception as e:
            self.logger.error(f"Failed to start Discord client: {e}")

    async def run_migration(self, discord_token: str, stoat_token: str, source_chan: int, target_chan: str):
        self.logger.info(f"Starting message migration from {source_chan} to {target_chan}")
        client = discord.Client(intents=discord.Intents.default())

        @client.event
        async def on_ready():
            try:
                channel = client.get_channel(source_chan)
                if not channel:
                    self.logger.error("Discord channel not found.")
                    return

                # Fetch all messages
                messages = []
                async for msg in channel.history(limit=None, oldest_first=True):
                    messages.append(msg)
                
                self.logger.info(f"Found {len(messages)} messages to migrate.")
                
                for msg in messages:
                    author_name = msg.author.display_name or msg.author.name
                    content = msg.clean_content
                    # Basic masquerade
                    masquerade = {"name": author_name, "avatar": str(msg.author.display_avatar.url)}
                    
                    if self.config.get('dry_run'):
                        self.logger.info(f"[DRY RUN] Would migrate message from {author_name}")
                        continue

                    await self.stoat_request(stoat_token, "POST", f"/channels/{target_chan}/messages", {
                        "content": content if content else "*(Attachment/Embed)*",
                        "masquerade": masquerade
                    })
                
                self.logger.info("Message migration finished.")
            except Exception as e:
                self.logger.error(f"Error in migration: {e}")
            finally:
                await client.close()

        await client.start(discord_token)

File: src/backend/engine.js
import { Client, GatewayIntentBits, PermissionsBitField } from 'discord.js';
import axios from 'axios';
import FormData from 'form-data';

export class MigrationEngine {
    constructor(config, logCallback) {
        this.config = config;
        this.logCallback = logCallback;
        this.stoatApi = 'https://api.stoat.chat';
        this.stoatCdn = null;
        this.roleMap = {};
        this.channelMap = {};
        this.messageAuthorCache = {};
        this.avatarCache = {};
    }

    async log(message) {
        const timestamp = new Date().toISOString();
        const formattedMsg = `${timestamp} - INFO - ${message}`;
        await this.logCallback(formattedMsg);
    }

    async getStoatCdnUrl(token) {
        if (this.stoatCdn) return this.stoatCdn;

        try {
            const response = await axios.get(this.stoatApi, { timeout: 5000 });
            const cdnUrl = response.data?.features?.autumn?.url;
            if (cdnUrl) {
                this.stoatCdn = cdnUrl.replace(/\/$/, '');
                await this.log(`Detected Stoat CDN URL: ${this.stoatCdn}`);
                return this.stoatCdn;
            }
        } catch (err) {
            await this.log(`Failed to fetch CDN URL from API root: ${err.message}`);
        }

        this.stoatCdn = 'https://cdn.stoatusercontent.com';
        return this.stoatCdn;
    }

    async uploadToStoat(token, fileUrl, filename, tag = 'attachments') {
        if (this.config.dry_run) {
            await this.log(`[DRY RUN] Would upload ${filename} to Stoat CDN (${tag})`);
            return 'dry-run-file-id';
        }

        const cdnBase = await this.getStoatCdnUrl(token);

        try {
            // Download from Discord CDN
            const fileResponse = await axios.get(fileUrl, { responseType: 'arraybuffer' });
            const fileData = Buffer.from(fileResponse.data);

            // Upload to Stoat CDN
            const form = new FormData();
            form.append('file', fileData, { filename });

            const response = await axios.post(`${cdnBase}/${tag}`, form, {
                headers: {
                    ...form.getHeaders(),
                    'X-Bot-Token': token
                }
            });

            return response.data.id;
        } catch (err) {
            await this.log(`Error uploading ${filename}: ${err.message}`);
            return null;
        }
    }

    async stoatRequest(token, method, path, jsonData = null) {
        const url = `${this.stoatApi}${path}`;
        const headers = { 'X-Bot-Token': token };

        for (let attempt = 0; attempt < 3; attempt++) {
            try {
                const config = { headers };
                if (jsonData) config.data = jsonData;

                const response = await axios({ method, url, ...config });

                if (response.status === 200 || response.status === 201) {
                    return response.data;
                }
            } catch (err) {
                if (err.response?.status === 429) {
                    const retryAfter = (err.response.data?.retry_after || 1000) / 1000;
                    await this.log(`Rate limited. Waiting ${retryAfter}s...`);
                    await new Promise(resolve => setTimeout(resolve, (retryAfter + 0.1) * 1000));
                    continue;
                } else if (err.response?.status >= 500) {
                    await this.log(`Stoat API Gateway Error (${err.response.status}). Retrying in 2s...`);
                    await new Promise(resolve => setTimeout(resolve, 2000));
                    continue;
                } else {
                    const error = err.response?.data || err.message;
                    await this.log(`Stoat API Error (${method} ${path}): ${err.response?.status} - ${error}`);
                    return null;
                }
            }
        }
        return null;
    }

    formatMessageContent(msg) {
        const parts = [];

        // 1. Handle Replies
        if (msg.reference && msg.reference.messageId) {
            const refId = msg.reference.messageId;
            const replyUser = this.messageAuthorCache[refId];
            parts.push(`> *Replying to ${replyUser || 'a message'}*`);
        }

        // 2. Rich Embeds (simplified - Discord.js handles embeds differently)
        if (msg.embeds && msg.embeds.length > 0) {
            for (const embed of msg.embeds) {
                if (embed.url && !msg.content?.includes(embed.url)) {
                    parts.push(embed.url);
                }
            }
        }

        // 3. Main Content
        if (msg.content) {
            parts.push(msg.content);
        }

        return parts.join('\n');
    }

    mapPermissions(discordPerms) {
        let stoatBits = 0;

        if (discordPerms.has(PermissionsBitField.Flags.ViewChannel)) stoatBits |= (1 << 20);
        if (discordPerms.has(PermissionsBitField.Flags.SendMessages)) stoatBits |= (1 << 22);
        if (discordPerms.has(PermissionsBitField.Flags.ManageMessages)) stoatBits |= (1 << 24);
        if (discordPerms.has(PermissionsBitField.Flags.ManageChannels)) stoatBits |= (1 << 2);
        if (discordPerms.has(PermissionsBitField.Flags.ManageRoles)) stoatBits |= (1 << 3);
        if (discordPerms.has(PermissionsBitField.Flags.Connect)) stoatBits |= (1 << 30);
        if (discordPerms.has(PermissionsBitField.Flags.Speak)) stoatBits |= (1 << 31);

        return stoatBits;
    }

    async getExistingStructure(token, serverId) {
        const server = await this.stoatRequest(token, 'GET', `/servers/${serverId}`);
        if (!server) return [{}, {}];

        const roles = {};
        if (server.roles) {
            for (const [rid, r] of Object.entries(server.roles)) {
                roles[r.name.toLowerCase()] = rid;
            }
        }

        const channelIds = server.channels || [];
        const channels = {};

        // Fetch channel details in parallel
        const channelPromises = channelIds.map(cid =>
            this.stoatRequest(token, 'GET', `/channels/${cid}`)
        );
        const results = await Promise.all(channelPromises);

        for (const c of results) {
            if (c && c.name) {
                channels[c.name.toLowerCase()] = c._id;
            }
        }

        return [roles, channels];
    }

    async runClone(discordToken, stoatToken, sourceId, targetId) {
        await this.log(`Starting cloning task for server ${sourceId} to ${targetId}`);

        const client = new Client({
            intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMembers]
        });

        return new Promise((resolve, reject) => {
            client.once('clientReady', async () => {
                try {
                    const guild = client.guilds.cache.get(sourceId.toString());
                    if (!guild) {
                        await this.log('Discord Guild not found.');
                        await client.destroy();
                        return resolve();
                    }

                    const [existingRoles, existingChannels] = await this.getExistingStructure(stoatToken, targetId);

                    // 1. Clone Roles
                    const sortedRoles = Array.from(guild.roles.cache.values())
                        .filter(r => !r.managed && r.id !== guild.id) // Skip @everyone and managed roles
                        .sort((a, b) => a.position - b.position);

                    for (const role of sortedRoles) {
                        const nameLower = role.name.toLowerCase();
                        if (existingRoles[nameLower]) {
                            this.roleMap[role.id] = existingRoles[nameLower];
                            continue;
                        }

                        if (this.config.dry_run) {
                            await this.log(`[DRY RUN] Would create role: ${role.name}`);
                            this.roleMap[role.id] = `dry_role_${role.id}`;
                            continue;
                        }

                        const res = await this.stoatRequest(stoatToken, 'POST', `/servers/${targetId}/roles`, {
                            name: role.name,
                            permissions: [this.mapPermissions(role.permissions), 0],
                            colour: role.color ? `#${role.color.toString(16).padStart(6, '0')}` : null,
                            hoist: role.hoist
                        });

                        if (res?.id) {
                            this.roleMap[role.id] = res.id;
                            await this.log(`Created role: ${role.name}`);
                        }
                    }

                    // 2. Clone Channels & Categories
                    const stoatCategories = [];
                    const categories = Array.from(guild.channels.cache.values())
                        .filter(c => c.type === 4) // Category type
                        .sort((a, b) => a.position - b.position);

                    for (const category of categories) {
                        const catChannels = [];
                        const children = Array.from(category.children.cache.values())
                            .sort((a, b) => a.position - b.position);

                        for (const channel of children) {
                            if (channel.type !== 0 && channel.type !== 2) continue; // Text or Voice

                            const nameLower = channel.name.toLowerCase();
                            if (existingChannels[nameLower]) {
                                const chanId = existingChannels[nameLower];
                                this.channelMap[channel.id] = chanId;
                                catChannels.push(chanId);
                                continue;
                            }

                            if (this.config.dry_run) {
                                await this.log(`[DRY RUN] Would create channel: ${channel.name} in ${category.name}`);
                                const chanId = `dry_chan_${channel.id}`;
                                this.channelMap[channel.id] = chanId;
                                catChannels.push(chanId);
                                continue;
                            }

                            const res = await this.stoatRequest(stoatToken, 'POST', `/servers/${targetId}/channels`, {
                                name: channel.name,
                                type: channel.type === 0 ? 'Text' : 'Voice',
                                description: channel.topic || null
                            });

                            if (res?._id) {
                                const chanId = res._id;
                                this.channelMap[channel.id] = chanId;
                                catChannels.push(chanId);
                                await this.log(`Created channel: ${channel.name}`);
                            }
                        }

                        if (catChannels.length > 0) {
                            stoatCategories.push({
                                id: Math.random().toString(36).substring(7),
                                title: category.name,
                                channels: catChannels
                            });
                        }
                    }

                    if (stoatCategories.length > 0 && !this.config.dry_run) {
                        await this.stoatRequest(stoatToken, 'PATCH', `/servers/${targetId}`, {
                            categories: stoatCategories
                        });
                    }

                    await this.log('Server structure cloning finished.');
                    await client.destroy();
                    resolve();
                } catch (err) {
                    await this.log(`Error in cloning: ${err.message}`);
                    await client.destroy();
                    reject(err);
                }
            });

            client.login(discordToken).catch(err => {
                this.log(`Failed to start Discord client: ${err.message}`);
                reject(err);
            });
        });
    }

    async runMigration(discordToken, stoatToken, sourceChan, targetChan, afterId = null) {
        await this.log(`Starting message migration from ${sourceChan} to ${targetChan}${afterId ? ' after ' + afterId : ''}`);

        const client = new Client({
            intents: [
                GatewayIntentBits.Guilds,
                GatewayIntentBits.GuildMessages,
                GatewayIntentBits.MessageContent
            ]
        });

        return new Promise((resolve, reject) => {
            client.once('clientReady', async () => {
                try {
                    const channel = client.channels.cache.get(sourceChan.toString());
                    if (!channel) {
                        await this.log('Discord channel not found.');
                        await client.destroy();
                        return resolve();
                    }

                    let messages = [];

                    // Fetch starting message if afterId provided
                    if (afterId) {
                        try {
                            const startingMsg = await channel.messages.fetch(afterId.toString());
                            messages.push(startingMsg);
                            await this.log(`Including starting message: ${afterId}`);
                        } catch {
                            await this.log(`Starting message ${afterId} not found or inaccessible.`);
                        }
                    }

                    // Fetch messages
                    const fetchOptions = { limit: 100 };
                    if (afterId) fetchOptions.after = afterId.toString();

                    let lastId = afterId?.toString();
                    while (true) {
                        const opts = { limit: 100 };
                        if (lastId) opts.after = lastId;

                        const fetched = await channel.messages.fetch(opts);
                        if (fetched.size === 0) break;

                        const sorted = Array.from(fetched.values()).sort((a, b) => a.createdTimestamp - b.createdTimestamp);
                        messages.push(...sorted);
                        lastId = sorted[sorted.length - 1].id;

                        if (fetched.size < 100) break;
                    }

                    await this.log(`Found ${messages.length} messages to migrate.`);

                    // Fetch threads
                    await this.log('Fetching threads from channel...');
                    const allThreads = [];

                    try {
                        // Active threads
                        const activeThreads = await channel.threads.fetchActive();
                        allThreads.push(...activeThreads.threads.values());
                        await this.log(`Found ${activeThreads.threads.size} active threads`);

                        // Archived threads
                        const archivedThreads = await channel.threads.fetchArchived({ fetchAll: true });
                        allThreads.push(...archivedThreads.threads.values());

                        await this.log(`Found ${allThreads.length} total threads (active + archived)`);

                        // Process each thread
                        for (const thread of allThreads) {
                            await this.log(`Processing thread: ${thread.name}`);
                            const threadMessages = [];

                            let lastThreadId = null;
                            while (true) {
                                const opts = { limit: 100 };
                                if (lastThreadId) opts.before = lastThreadId;

                                const fetched = await thread.messages.fetch(opts);
                                if (fetched.size === 0) break;

                                const sorted = Array.from(fetched.values()).sort((a, b) => a.createdTimestamp - b.createdTimestamp);
                                threadMessages.push(...sorted);
                                lastThreadId = sorted[0].id;

                                if (fetched.size < 100) break;
                            }

                            if (threadMessages.length > 0) {
                                // Thread start marker
                                const startMarker = {
                                    content: `[THREAD: '${thread.name}']`,
                                    createdTimestamp: thread.createdTimestamp,
                                    author: { displayName: 'System', username: 'System', displayAvatarURL: () => '' },
                                    attachments: new Map(),
                                    embeds: [],
                                    reference: null,
                                    id: '0',
                                    isThreadMarker: true,
                                    isStart: true
                                };
                                messages.push(startMarker);

                                messages.push(...threadMessages);

                                // Thread end marker
                                const endMarker = {
                                    content: '[end of THREAD]',
                                    createdTimestamp: threadMessages[threadMessages.length - 1].createdTimestamp,
                                    author: { displayName: 'System', username: 'System', displayAvatarURL: () => '' },
                                    attachments: new Map(),
                                    embeds: [],
                                    reference: null,
                                    id: '0',
                                    isThreadMarker: true,
                                    isStart: false
                                };
                                messages.push(endMarker);

                                await this.log(`Added ${threadMessages.length} messages from thread '${thread.name}'`);
                            }
                        }
                    } catch (err) {
                        await this.log(`Error fetching threads: ${err.message}`);
                    }

                    // Sort all messages by timestamp
                    messages.sort((a, b) => a.createdTimestamp - b.createdTimestamp);
                    await this.log(`Total messages to migrate (including threads): ${messages.length}`);
                    const total = messages.length;

                    let migratedCount = 0;
                    let firstMessageLink = null;

                    for (let idx = 0; idx < messages.length; idx++) {
                        const msg = messages[idx];

                        if (idx === 0 && msg.url) {
                            firstMessageLink = msg.url;
                        }

                        // Handle thread markers
                        if (msg.isThreadMarker) {
                            if (!this.config.dry_run) {
                                await this.stoatRequest(stoatToken, 'POST', `/channels/${targetChan}/messages`, {
                                    content: msg.content
                                });
                            }
                            await this.log(`PROGRESS:${idx + 1}/${total}`);
                            await this.log(`Sent thread marker: ${msg.content}`);
                            migratedCount++;
                            continue;
                        }

                        const authorName = msg.author.displayName || msg.author.username;
                        this.messageAuthorCache[msg.id] = authorName;

                        const formattedContent = this.formatMessageContent(msg);

                        // Handle attachments
                        const stoatAttachments = [];
                        if (msg.attachments && msg.attachments.size > 0) {
                            for (const [, att] of msg.attachments) {
                                const fileId = await this.uploadToStoat(stoatToken, att.url, att.name);
                                if (fileId) stoatAttachments.push(fileId);
                            }
                        }

                        // Generate timestamp and header
                        const timestamp = new Date(msg.createdTimestamp).toISOString().replace('T', ' ').replace(/\.\d{3}Z/, ' UTC');
                        const header = `*${timestamp}*\n`;

                        let finalContent;
                        if (!formattedContent) {
                            if (stoatAttachments.length > 0) {
                                finalContent = header;
                            } else {
                                finalContent = header + '*(Attachment/Embed)*';
                            }
                        } else {
                            finalContent = header + formattedContent;
                        }

                        const masquerade = {
                            name: authorName.substring(0, 32),
                            avatar: msg.author.displayAvatarURL()
                        };

                        if (this.config.dry_run) {
                            await this.log(`PROGRESS:${idx + 1}/${total}`);
                            await this.log(`[DRY RUN] Would migrate message from ${authorName}`);
                            migratedCount++;
                            continue;
                        }

                        const payload = {
                            content: finalContent,
                            masquerade
                        };
                        if (stoatAttachments.length > 0) {
                            payload.attachments = stoatAttachments;
                        }

                        await this.stoatRequest(stoatToken, 'POST', `/channels/${targetChan}/messages`, payload);
                        migratedCount++;
                        await this.log(`PROGRESS:${idx + 1}/${total}`);
                    }

                    // Send summary message
                    if (migratedCount > 0) {
                        const summaryText = `[Discord Terminator](<https://github.com/rambros3d/discord-terminator>) has migrated **${migratedCount}** messages from #${channel.name}${firstMessageLink ? ` ([link](<${firstMessageLink}>))` : ''}`;
                        await this.stoatRequest(stoatToken, 'POST', `/channels/${targetChan}/messages`, { content: summaryText });
                    }

                    await this.log('Message migration finished.');
                    await this.log('TASK_COMPLETE');
                    await client.destroy();
                    resolve();
                } catch (err) {
                    await this.log(`Error in migration: ${err.message}`);
                    await this.log('TASK_FAILED');
                    await client.destroy();
                    reject(err);
                }
            });

            client.login(discordToken).catch(err => {
                this.log(`Failed to start Discord client: ${err.message}`);
                reject(err);
            });
        });
    }
}
